plugins/optimization-detective/storage/rest-api.php

```diff
@@ -186,38 +186,84 @@ function od_handle_rest_request( WP_REST_Request $request ) {
 	OD_Storage_Lock::set_lock();
 
 	try {
-		// The "strict" URL Metric class is being used here to ensure additionalProperties of all objects are disallowed.
-		$url_metric = new OD_Strict_URL_Metric(
-			array_merge(
-				$data,
-				array(
-					// Now supply the readonly args which were omitted from the REST API params due to being `readonly`.
-					'timestamp' => microtime( true ),
-					'uuid'      => wp_generate_uuid4(),
-					'etag'      => $request->get_param( 'current_etag' ),
-				)
+		$data = array_merge(
+			$data,
+			array(
+				// Now supply the readonly args which were omitted from the REST API params due to being `readonly`.
+				'timestamp' => microtime( true ),
+				'uuid'      => wp_generate_uuid4(),
+				'etag'      => $request->get_param( 'current_etag' ),
 			)
 		);
-	} catch ( OD_Data_Validation_Exception $e ) {
-		return new WP_Error(
-			'rest_invalid_param',
-			sprintf(
+
+		/**
+		 * Filters the URL Metric data prior to validation for storage.
+		 *
+		 * This allows for custom sanitization and validation constraints to be applied beyond what can be expressed in
+		 * JSON Schema. This is also necessary because the 'validate_callback' key in a JSON Schema is not respected when
+		 * gathering the REST API endpoint args via the {@see rest_get_endpoint_args_for_schema()} function. Besides this,
+		 * the REST API doesn't support 'validate_callback' for any nested arguments in any case, meaning that custom
+		 * constraints would be able to be applied to multidimensional objects, such as the items inside 'elements'.
+		 *
+		 * This filter only applies when storing a URL Metric via the REST API. It does not run when a stored URL Metric is
+		 * loaded from the od_url_metrics post type. This means that sanitization and validation logic enforced via this
+		 * filter can be more expensive, such as doing filesystem checks or HTTP requests.
+		 *
+		 * To fail validation for the provided URL Metric data, an OD_Data_Validation_Exception exception should be thrown.
+		 * Otherwise, any filter callbacks must return an array consisting of the sanitized URL Metric data.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param array<string, mixed> $data URL Metric data. This is the Data type from OD_URL_Metric.
+		 */
+		$data = apply_filters( 'od_url_metric_data_pre_storage', $data );
+
+		// The "strict" URL Metric class is being used here to ensure additionalProperties of all objects are disallowed.
+		$url_metric = new OD_Strict_URL_Metric( $data );
```

Comment:
Can we not keep filtering the `OD_URL_Metric` instance for this, as that would be a better filter interface than dealing with raw data?

Reply:
Also, by passing around an object then the user needs to be aware of the distinction between `OD_Strict_URL_Metric` and `OD_URL_Metric`, where the former is intended to be an internal class used just in the REST API endpoint.

Reply:
That's only for the specific example of unsetting data though (because there's no `unset()` method). Regular sanitization would be as simple as:
```php
function ( $url_metric ) {
    $url_metric->set( 'prop', 'sanitized-value' );
    return $new_url_metric;
}
```

Honestly, at that point, it would be even better to use an action hook, since the object can be modified directly either way, and an exception can still be thrown too, so a filter is not needed. This would also be safer, as a filter would allow to exchange the entire object which, despite me having suggested that earlier in favor of `unset()`, also has drawbacks as normally people shouldn't be doing that.

If you feel strongly about the ergonomics of this for the scenario of unsetting a property, I'd rather bring back the `unset()` method than filtering raw data just because of the lack of an `unset()` method.

Reply:
There is no `set()` method either. So you'd have to do the same:

```php
function ( $url_metric ) {
    $data = $url_metric->jsonSerialize();
    $data['prop'] = 'sanitized-value';
    $new_url_metric = new OD_Strict_URL_Metric( $data );
    return $new_url_metric;
}
```

If we were to support `set()` then this would bring back concerns about cache invalidation for the group and the collection, and it would be confusing if a user tries to call `set()` on a URL Metric from a collection in a tag visitor, as obviously that would not be right since it's only relevant here.

So I think it's preferable for the `OD_URL_Metric` object to be immutable, siding with you for suggesting that we don't allow `unset()`.

Reply:
Oh I must have confused seeing `OD_Element::offsetSet()`.

Taking a step back: The issues of passing around raw array data are that there is little control over it, no way to insert any kind of "middleware", plus no good property autocomplete support etc. So I think it's worth finding a better solution, and the current constraints of `OD_URL_Metric` and `OD_Strict_URL_Metric` don't seem right, based on the points you made.

How about something like this?
* We introduce another kind of URL metric class. This class would represent a URL metric without any validation in place and would only be used while a URL metric is being recorded.
* This class would have a set method, and could also have an unset method, as modifying what's recorded is crucial as part of what we're trying to achieve here. It would be passed to the new action (or filter) and allow ergonomically to customize what's being recorded, without any constraints of schema validation.
* After the hook, an instance of the current URL metric class is created from the "unvalidated" instance, so at that point validation fires and would catch any wrongdoing that might have happened during the hook.

A few architectural alternatives on how this could be implemented:
1. We introduce something like `OD_Unvalidated_URL_Metric` and have `OD_URL_Metric` extend it. Downside: A bit odd naming pattern with the negation of "validated" in the base class.
2. We name the new unvalidated class `OD_URL_Metric` and rename the current `OD_URL_Metric` to something like `OD_Validated_URL_Metric`, and have it extend the new `OD_URL_Metric`. Downside: A bit of a strange pattern would be required to override the set and unset methods to throw an exception instead of actually setting - not great to have methods on a class that shouldn't be there.
3. We introduce the new class separately from the existing URL metric classes, i.e. no new class extension. This way only the new class would have set and unset methods. Maybe this could be called something like `OD_Raw_URL_Metric`? Then either that class would have a method to return a validated URL metric instance from its data, or the current `OD_URL_Metric` would have a static method like `create_from_raw()` (or similar).
    * For this approach, ideally there should be an interface that both the new `OD_Raw_URL_Metric` and the current `OD_URL_Metric` would implement, to make sure the common part of their implementations remain aligned. If we go with an interface, maybe the interface itself should be named `OD_URL_Metric` and the current `OD_URL_Metric` class should then be renamed to something more specific, like `OD_Validated_URL_Metric`.

If we go with the interface or a renaming of the current `OD_URL_Metric` class, I realize this would be a bit of refactoring, so not ideal to do so shortly before the release as part of this PR. But once we agree on an approach, we could in this PR simply do the bare minimum to work towards it and address the rest in a follow up. For instance, if we decide to rename `OD_URL_Metric`, we don't have to do it as part of this PR. The main thing IMO is that we can discuss and decide on an option here, without thinking about the constraints of this PR too much, and then do the least effort needed towards that goal in this PR.

Reply:
> The issues of passing around raw array data are that there is little control over it, no way to insert any kind of "middleware", plus no good property autocomplete support etc.

The filters are the middleware. Property auto completion is not currently supported anyway for custom properties. Only core properties could be possible autocompleted, and this can be implemented by a plugin importing the PHPStan `Data` type from  the `OD_URL_Metric` class.

> How about something like this?

I really think this is too much. I think it is somewhat over-engineered. What is in place right now feels WordPressy with a simple filter to process the raw data (which would have been pre-validated with the JSON schema attached to the REST API args), and then will be validated again anyway when the `OD_Strict_URL_Metric` is constructed. The filter in place now is similar to the `wp_insert_post_data` filter in WordPress so it's easy to understand and use.

If we want something more structured in the future I suggest deprecating this filter to instead go with the replacement. But I really think this current approach is best, especially for now.

Reply:
> The filters are the middleware.

Not in the sense that I meant it, hey don't allow controlling what you can modify. With an array you can do whatever you want. If we later renamed a property, or wanted to trigger a warning if someone changes a specific property, things like that, we can't do that if we use an array - but we can if we use a class instance.

> I really think this is too much. I think it is somewhat over-engineered.

Can you clarify why? What's "WordPressy" is a bit subjective, and FWIW some parts of WordPress are not necessarily examples for good engineering.

> If we want something more structured in the future I suggest deprecating this filter to instead go with the replacement.

That's a short-term option, though it would not be a good long-term solution IMO per my previous comment. Since this is in the milestones due tomorrow, I'd be okay going with it for now, but it's really not great to ship new API surfaces without sufficient consideration just because we want to ship something fast.

Reply:
> If we later renamed a property and handle it backwards compatibly, or wanted to trigger a warning if someone changes a specific property, things like that, we can't do that if we use an array - but we can if we use a class instance.

That's a good point. 

> Can you clarify why?

It's subjective 😀

> What's "WordPressy" is a bit subjective

I'm not sure its subjective. WordPressy is what aligns with previous established patterns in WordPress.

> FWIW some parts of WordPress are not necessarily examples for good engineering.

This is true.

----

I think there is another way to go about this with an existing filter in the REST API without introducing a new sanitization/validation mechanism: the `rest_request_brefore_callbacks` filter. I've never used this filter before, but it is billed as being specifically for performing additional validation. I'm investigating.

Reply:
Here, check this out: d4c9f40fb3307d11b1679da76c2e8333a80f91d0

Reply:
Sure, if we can implement this without a new API surface in OD entirely, that would simplify things, at least for now :)

Reply:
Thanks, that looks great and elegantly avoids expanding the new API. So obviously this is a workaround now, but it gives us time to think whether and how we would want to make this more intuitively possible.